### PR TITLE
Swaps: Asset to Pay / Asset to Receive

### LINKED ISSUE #20332: Asset to Receive selection drawer (From Setup Swap screen)
image

https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=43-170640&m=dev

Depends on https://github.com/status-im/status-mobile/issues/20380

### LINKED ISSUE #20333: Asset to Pay selection drawer (From Setup Swap screen)
Implement Asset to Pay selection drawer

https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=76-130105&t=DgFnr1o1B4N17NwC-1

image

Dependency #20380

### PR DESCRIPTION
**Pull request by alwx**
fixes #20332
fixes #20333

### Summary

Adds sheets for both "Asset to pay" and "Asset to receive"

#### Platforms

- Android
- iOS

status: ready

### CODE REVIEW

**Comment by ulisesmac** [comment]
Hi @alwx 

It'd be great if you add a section with the steps to test and also some demo screenshots or video of it working to the description. It'll also help QA to test this PR

There has been a lot of files changed, so any additional note on the code will also help. :+1: 

Thank you!

**Comment by briansztamfater** [comment]
If it depends on #21139 maybe would be better to use #21139 branch (`feature/20380`) as the base branch so we can review the relevant changes in this PR

File: src/status_im/contexts/wallet/sheets/select_asset/asset_list/style.cljs
Comment (briansztamfater) [inline src/status_im/contexts/wallet/sheets/select_asset/asset_list/style.cljs:1]: Seems this file is not needed

File: src/status_im/subs/wallet/wallet.cljs
Comment (briansztamfater) [inline src/status_im/subs/wallet/wallet.cljs]: Can we add more context of this TODO? Maybe add a GitHub issue to keep track of follow ups?

**Comment by briansztamfater** [comment]
Does this PR includes integration within the swap flow or just the screens?
In case is just the screens, maybe add a way to open them so QA can test them in isolation, and add intruction in the PR description.

**Comment by alwx** [comment]
@briansztamfater @status-im/mobile-qa the integration is done as well. This is how it works:

https://github.com/user-attachments/assets/fe09a495-0404-47e9-a311-77903a6dbd93

**Review by briansztamfater — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
Hi @alwx, could you please resolve the existing conflicts and rebase the current PR if it's ready for testing? Thanks!

**Comment by alwx** [comment]
@VolodLytvynenko seems like this PR has already been tested in https://github.com/status-im/status-mobile/pull/21269 because that PR is based on this one?

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko seems like this PR has already been tested in #21269 because that PR is based on this one?

@alwx got it. Thanx for clarification

**Comment by alwx** [comment]
@VolodLytvynenko But did the rebase

**Comment by VolodLytvynenko** [comment]
Hi @alwx thank you for PR. Take a look found issues.

## ISSUE 1: Unsupported tokens by network are shown in the "Asset to Pay" / "Asset to Receive" list
### Steps:
1. Restore a user account that holds tokens unsupported on certain networks but available on several networks.
2. Select a token available on multiple networks (e.g., ETH).
3. Choose the Optimism or Arbitrum network from the available options in the drawer.
4. Navigate to the swap page.
5. Open the "Asset to Pay" or "Asset to Receive" list.
6. Check whether unsupported tokens for the current network are displayed 


### Actual Result:
Unsupported tokens for the selected network are shown in the "Asset to Pay" / "Asset to Receive" list, this leads to that we can prosseds

e.g., Arbitrum token not supported on Optimism network
Optimism token not supported on Arbitrum network

https://github.com/user-attachments/assets/9306ab2c-3889-430d-b776-576c735d0e8a

### Expected Result:
Only tokens supported by the selected network should be displayed in the "Asset to Pay" / "Asset to Receive" list.

**Comment by VolodLytvynenko** [comment]
## ISSUE 2 Blinks occur after tapping the search field
### Steps:
1. Go to the swap page.
2. Open "Select asset to pay" or "Asset to receive."
3. Tap the search field.

### Actual result:
- On Android, the screen blinks after tapping the search asset field.

https://github.com/user-attachments/assets/cddf72b0-c579-4a68-a55a-14775b930bf8

- On iOS, the keyboard does not open smoothly after tapping the search asset field.

https://github.com/user-attachments/assets/ba1efedc-1bb9-4527-ab92-32057bda8245

### Expected result:
No blinks or on Android.
The keyboard should open smoothly on iOS without delay.

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
## ISSUE 3: The  'Approve' button is shown, inspite the approve button is not related to certain token

### Steps:
1. Go to swap page
2. Select any ERC-20 non-approved token (e.g. DAI)
3. Enter some value (e.g. 1 DAI)
4. Open asset to pay screen
5. Select any other token (e.g. ETH or token which already approved)

### Actual result:
- The  'Approve' button is shown, inspite the approve button is not related to ETH token, this button appears only for ERC-20 tokens

https://github.com/user-attachments/assets/8896a77a-de43-4232-9198-dca68292c57d

### Expected result:
After selecting a new token, the "Approve" button should appear for tokens only for those that require approval

**Comment by VolodLytvynenko** [comment]
I believe this issue is related to [issue 3](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2361541243)

## ISSUE 4: "Review Swap" button is enabled after selection new token within Asset to Pay screen even when the user lacks sufficient tokens to perform the swap

### Steps:
1. Restore a user account with at least 2 tokens.
2. Navigate to the swap page.
3. Select an ERC-20 approved token (e.g., DAI).
4. Enter a valid available amount (e.g., 1 DAI).
5. Open the "Asset to Pay" screen.
6. Select a token that the user does not have enough balance for (e.g., Wrapped Bitcoin or ETH).

### Actual result:
The "Review Swap" button remains enabled, allowing the user to proceed with the swap, even though there aren't enough tokens for the transaction.

https://github.com/user-attachments/assets/0e49e444-339d-4238-9d36-650312f1cebb

### Expected result:
The "Review Swap" button should be disabled if the user doesn't have sufficient assets for the newly selected token.

### Devices:
Pixel 7a, Android 13
iPhone 11 Pro Max, iOS 17

File: src/status_im/contexts/wallet/swap/setup_swap/view.cljs
Comment (briansztamfater) [inline src/status_im/contexts/wallet/swap/setup_swap/view.cljs:289]: I'd rather use a modal with sheet presentation instead of a bottom sheet. Check the implementation of set spending cap screen.

https://github.com/status-im/status-mobile/blob/0eb065bbb723be6a235d41b14fb28bf2a65a9ba2/src/status_im/navigation/screens.cljs#L543-L545

https://github.com/status-im/status-mobile/blob/052e412c60efa80e062e747a61d2e2454070882d/src/status_im/contexts/wallet/swap/setup_swap/view.cljs#L383

**Comment by VolodLytvynenko** [comment]
### ISSUE 5: The default tokens are shown as disabled on "Asset to Pay" page if those tokens are with 0 value

### Steps:
1. Create new account or restore an account without any assets
2. Go to "Asset to Pay" page

### Actual result:
Default tokens like SNT, DAI, and ETH are shown as disabled in the "Your assets on Mainnet" section if they have a 0 balance.
image

### Expected result:
Tokens with a 0 balance should not be displayed. Only tokens with available assets should be shown.image

**Comment by VolodLytvynenko** [comment]
## PR_ISSUE 6: "Popular assets on "network" " section is not shown within "Asset to Pay" sheet

### Steps:
1. Open Asset to Pay sheet

### Actual result:
image

### Expected result:
Popular assets on <network> sheet is shown on "Asset to Pay" sheet
image

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
## ISSUE 7: The token is not disabled in the "Asset to Pay" sheet if the same token is already selected in the "Receive" section
### Steps:
1. Go to the swap page.
2. Select a token for the "Receive" section (e.g., ETH).
3. Open the "Asset to Pay" sheet.

### Actual result:
The same token (e.g., ETH) is not disabled in the "Asset to Pay" sheet even though it is already selected in the "Receive" section.

https://github.com/user-attachments/assets/ec6759ea-e3ac-4b0c-8b45-40d5ea03a721

### Expected result:
The token selected in the "Receive" section should be disabled in the "Asset to Pay" sheet to prevent swaping on the same token

How it works on a desktop:

https://github.com/user-attachments/assets/ac33b76e-4fef-463d-818d-7ca68f8560eb

**Comment by alwx** [comment]
@VolodLytvynenko ready to be re-checked

**Comment by VolodLytvynenko** [comment]
Hi @alwx thank you for fixes. Take a look found issues:

[Issue 1](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2361427598)  is not fully resolved yet. Tokens not supported by the current network can still be found and selected. However, our validation prevents transactions from proceeding with these tokens, so this issue can be considered a follow-up. 


### Steps:
1. Navigate to the swap page.
2. Open the "Asset to Pay" or "Asset to Receive" list.
3. Search for unsupported tokens (example: if network is Optimism, then search for Arb token)
4. Check whether unsupported tokens for the current network are displayed

### Actual Result:
Unsupported tokens for the selected network are shown in the "Asset to Pay" / "Asset to Receive" list

https://github.com/user-attachments/assets/db40e1c2-5333-464e-a4c8-f7093f1da436

**Comment by VolodLytvynenko** [comment]
@alwx Apologies for message https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373767725. Ignore it please. Issue is fixed

**Comment by VolodLytvynenko** [comment]
[Issue 2](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2361428621) is not fixed yet; the blinks are still shown. However, this can be considered a follow-up.

- Pixel 7a, Android 13

https://github.com/user-attachments/assets/c19bd477-d0c8-498f-b2de-7ba69b8d588c

- iPhone 11 Pro Max, IOS 17

https://github.com/user-attachments/assets/8f564a5a-155c-4308-9957-a0cd3a98e430

**Comment by VolodLytvynenko** [comment]
[issue 3](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2361541243) is fixed

**Comment by VolodLytvynenko** [comment]
## ISSUE 8: Receive value is not recalculated correctly when new token is selected within Asset to Receive sheet until autorefresh

### Steps:
1. Go swap page
2. Prefulfill into Send field
3. Change token in recieve value
4. Check recalculation in receive value

### Actual result:
The value in the Receive field is incorrectly calculated. For example:
 - Initial conversion: DAI 1 -> ETH 0.00037
 - After changing ETH to USDC, it shows: DAI 0.01 -> USDC 378915182.67 instead of DAI 0.01 -> USDC 0.0096.

https://github.com/user-attachments/assets/06344560-5591-4abd-b1b8-775bed3af093

### Expected result:
When the token is changed, the data should be recalculated and displayed correctly. An update or upload statement should be shown reflecting the new calculation.

**Comment by VolodLytvynenko** [comment]
[issue 4](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2361593502) is fixed

**Comment by VolodLytvynenko** [comment]
Current issue is with very low priority and can be considered in 2.32 as follow up
 
[issue 5](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2367881461)  is partially fixed. The default tokens are no longer shown as disabled, which is resolved. However, when switching to Network 2, tokens like SNT, DAI, and ETH that exist on Network 1 are shown with a value of 0 on the "Your assets on Mainnet" section, even though they are not present on Network 2


### Steps:
1. Create new account or restore an account with snt, dai or eth only on Network 1
2. Select Network 2 for swapping on which tokens are not present, but present on Network 1
3. Go to "Asset to Pay" page

### Actual result:
Default tokens (SNT, DAI, ETH) are displayed with a 0 value in the "Your assets on Mainnet" section, even though they are available on Network 1 but not on Network 2, while Network 2 is selected.

image

### Expected result:
Tokens should not be shown in "Your assets on Mainnet" section if they are not available on the currently selected network. 
https://figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=76-130105&node-type=frame&t=mccNGM33NrJBogCP-0

**Comment by VolodLytvynenko** [comment]
[issue 6](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2367893581) is fixed

**Comment by VolodLytvynenko** [comment]
[issue 7](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2367927093) is fixed

**Comment by VolodLytvynenko** [comment]
@alwx Great job! This PR is complex and has a lot of details. Appreciate your effort in handling it.

Just a quick summary of the remaining issues:
[Issue 8](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373894888) needs to be addressed in current PR
[Issue 2](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2361428621) would be nice to fix within this PR, but it's fine to handle separately if you prefer.
[Issue 5](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2374068702) will be created as a follow-up.

**Comment by alwx** [comment]
@VolodLytvynenko issue 8 should be fixed in the last commit (pushed it ~30 mins ago). Could you recheck?

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko issue 8 should be fixed in the last commit (pushed it ~30 mins ago). Could you recheck?

@alwx the issue is still reproducible for me. The data still wrong calculated in the receive field

https://github.com/user-attachments/assets/95f68675-2a98-4484-8bd4-71da833faf6f

**Comment by VolodLytvynenko** [comment]
Additional issue 9, which I believe also related to [issue 8](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373894888)

The error message is shown which is related to previous swap after the token is changed in the receive field

### Steps:
1. Set up swap tokens to display an error (example: Send 1 DAI -> Receive SNT will cause the 'not enough liquidity' error message)
2. Change asset SNT to luqide one (example to ETH)

### Actual result:
The error message related to a previous swap attempt (e.g., "not enough liquidity") is still displayed even after the token in the receive field is changed to a liquid asset like ETH.

https://github.com/user-attachments/assets/d8e96bd4-bf15-4ddc-a447-ba57471a3c87

### Expected result:
The error message should be cleared once the user changes the token in the receive field suggesting  to user to proceed with the swap flow

**Comment by VolodLytvynenko** [comment]
@alwx Last time, I thought [issue 1](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373767725) was my mistake, possibly due to mixing up some builds. However, it's now consistently reproducible on my end. Although it might not happen frequently, hopefully the logs will help. Could you take a look as well?

As shown in the screenshot, while I'm on the Optimism network, I'm able to select non-Optimism tokens like ARB or Aragon in both the "Asset to Pay" and "Asset to Receive" sheets.

https://github.com/user-attachments/assets/b4ef1df5-371f-4b44-bf16-be3dadd26c12

### Logs:
[logs (2).zip](https://github.com/user-attachments/files/17133393/logs.2.zip)

**Comment by alwx** [comment]
@VolodLytvynenko I got it — it's literally a rebase issue. Fixed in the last commit, please re-check! (both issue 8 and 9)

**Comment by VolodLytvynenko** [comment]
Unfortunately one additional issue is found

## ISSUE 10: "Popular Assets" section is not shown in "Asset to Pay" and "Popular Assets" sheets until relogin
### Steps:
1. Recover the user
2. Check "Popular Assets" and "Popular Assets" sheets

### Actual result:
 "Popular Assets" section is not fetched in "Asset to Pay" and "Popular Assets" sheets until relogin

https://github.com/user-attachments/assets/5afce5dc-1b49-4e10-aa02-fcdd9fb6b6ce

### Expected result:
 "Popular Assets" section is fetched in "Asset to Pay" and "Popular Assets" sheets

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### Logs:
[Status-debug-logs (1).zip](https://github.com/user-attachments/files/17134217/Status-debug-logs.1.zip)

**Comment by VolodLytvynenko** [comment]
[issue 8](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373894888) is not fixed properly.

When changing the token in the Receive field on the swap page, the recalculation does not occur, and additionally the 'Approve' button disappears for non-approved ERC-20 tokens.

### Steps:
1. Go swap page
2. Prefulfill "Send" field
3. Change token in receive value
4. Check recalculation in receive value

### Actual result:
The value in the Receive field is shown as 0 after token is changed

https://github.com/user-attachments/assets/4eaac3e5-11c7-4c75-9c94-86d75ea0149d

### Expected Result:
- The Receive field should be recalculated correctly when the token is changed.
- The 'Approve' button should remain visible for non-approved ERC-20 tokens.

**Comment by VolodLytvynenko** [comment]
[issue 1](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373767725) mentioned in [recent comment](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2374346508) is also consistently reproduced now

**Comment by alwx** [comment]
@VolodLytvynenko From what it seems issue 10 is reproducable because of the feature flag. You restore the user so the feature flag is not yet set.

**Comment by alwx** [comment]
@VolodLytvynenko ready now. Please, check issue 8, 10 and 1 — all of them should be fixed now.

**Comment by VolodLytvynenko** [comment]
hi @alwx 
[issue 8](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373894888) still happens occasionally, especially on iOS. The recalculation isn't smooth (the wrong recalculated value is blinked), and sometimes it blinks with the "something went wrong" validation

- Pixel 7a, Android 13

https://github.com/user-attachments/assets/5172d973-3765-470a-bc04-1644d1c2164a

- iPhone 11 Pro max, IOS 17

https://github.com/user-attachments/assets/35f97c1c-b80f-4321-8b4f-ed4b9fb3f762


### Logs:
Android logs
[logs.zip](https://github.com/user-attachments/files/17151877/logs.zip)


IOS logs
[logs.zip](https://github.com/user-attachments/files/17151862/logs.zip)

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko From what it seems issue 10 is reproducable because of the feature flag. You restore the user so the feature flag is not yet set.

@alwx Perhaps I didn't understand you here. [Issue 10](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2374449861) is still reproducible for both recovered and new users after swap feature flag is enabled, and popular assets section is not shown until the user logs out and logs back in

https://github.com/user-attachments/assets/b7572ce1-5d4c-4f34-b3c3-9f570e0e0bdb

### logs:
[logs (2).zip](https://github.com/user-attachments/files/17152135/logs.2.zip)

**Comment by VolodLytvynenko** [comment]
issue 1 is fixed

**Comment by alwx** [comment]
@VolodLytvynenko issue 10 is fixed (I don't think it's an issue though because the loading happens on initialization — which is exactly what we needed because there won't be any feature flag in the future)

issue 8 should also be fixed — I tried different approach to loading now but i'm not sure it will work in 100% of cases because there are too many things that could go wrong on that particular screen

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko issue 10 is fixed (I don't think it's an issue though because the loading happens on initialization — which is exactly what we needed because there won't be any feature flag in the future)


Hi @alwx, thanks for the fixes! Take a look please one additional issue, now the Optimism popular assets section is limited to only 4 tokens, and I haven’t been able to see more than that, even after some relogins 

## ISSUE 11: Optimism Assets Not Fully Fetched in "Popular Assets" Section

### Steps to Reproduce:
1. Go to the swap page and select the Optimism network.
2. Check the "Popular Assets" section in the "Asset to Pay" or "Asset to Receive" sheets.

### Actual Result:
Only 4 tokens are shown in the "Popular Assets" section for Optimism.

https://github.com/user-attachments/assets/1a26a6c4-72b6-4c92-8584-35e459ef14bb


### Expected Result:
More popular tokens should be displayed in the "Popular Assets" section for the Optimism network, as on Desktop

https://github.com/user-attachments/assets/d7f98098-2c2f-44a6-9cc5-2cc5914c88ca


### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### Logs:
[logs.zip](https://github.com/user-attachments/files/17161776/logs.zip)

**Comment by VolodLytvynenko** [comment]
> > @VolodLytvynenko issue 10 is fixed (I don't think it's an issue though because the loading happens on initialization — which is exactly what we needed because there won't be any feature flag in the future)
> 
> Hi @alwx, thanks for the fixes! Take a look please one additional issue, now the Optimism popular assets section is limited to only 4 tokens, and I haven’t been able to see more than that, even after some relogins
> 
> ### Steps to Reproduce:
> 1. Go to the swap page and select the Optimism network.
> 2. Check the "Popular Assets" section in the "Asset to Pay" or "Asset to Receive" sheets.
> 
> ### Actual Result:
> Only 4 tokens are shown in the "Popular Assets" section for Optimism.
> 
>  4tokens.mp4 
> ### Expected Result:
> More popular tokens should be displayed in the "Popular Assets" section for the Optimism network, as on Desktop
> 
>  swap.mp4 
> ### Devices:
> * Pixel 7a, Android 13
> * iPhone 11 Pro Max, IOS 17
> 
> ### Logs:
> [logs.zip](https://github.com/user-attachments/files/17161776/logs.zip)

@alwx Better to fix is in a follow up. WDYT?

**Comment by alwx** [comment]
@VolodLytvynenko agree. That would make the whole process of fixing easier since it's a big PR.

Regarding 4 tokens for Optimism: I've just checked and that's actually what server (`status-go` side) returns us.

**Comment by VolodLytvynenko** [comment]
@alwx, thanks for fixing all the issues. This was a challenging feature, but I think our swap now looks even better than in some popular dapps! However, [issue 8](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2373894888) still happens occasionally, though rarely. The important remaining issue is [issue 11](https://github.com/status-im/status-mobile/pull/21140#issuecomment-2378811726), which needs to be addressed in 2.31. If you think it's better to fix it in this PR, let me know, and I'll recheck it; otherwise, the PR can be merged.

**Comment by VolodLytvynenko** [comment]
> Regarding 4 tokens for Optimism: I've just checked and that's actually what server (`status-go` side) returns us.

@alwx got it. Thank you for clarification

**Pull request merged**
